infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,56 +81,76 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheConflictKeys.get(conflictKey);
+    }
+
+    /**
+     * The first address in the priority queue.
+     */
+    public long firstAddress() {
+        if (cacheEntries.isEmpty())
+            return Address.NOT_FOUND;
+
+        return cacheEntries.peek().txVersion;
+    }
+
+    /**
+     * Invalidate the records with the minAddress. It could be one or multiple records
+     */
+    private void invalidateFirst() {
+        ConflictTxStream firstEntry = cacheEntries.peek();
+        if (firstEntry == null) {
+            return;
+        }
+
+        int i = 0;
+        while (!cacheEntries.isEmpty() && cacheEntries.peek().txVersion == firstEntry.txVersion) {
+            ConflictTxStream entry = cacheEntries.poll();
+            cacheConflictKeys.remove(entry);
+            i++;
+        }
+
+        log.debug("Evict {} entries", i);
+        maxConflictWildcard = Math.max(maxConflictWildcard, firstEntry.txVersion);
     }
 
     /**
-     * Invalidate all records up to a trim mark.
+     * Invalidate all records up to a trim mark (not included).
      *
      * @param trimMark trim mark
      */
     public void invalidateUpTo(long trimMark) {
         log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
+        int entries = 0;
 
-        AtomicLong entries = new AtomicLong();
-
-        conflictCache.asMap().forEach((key, txVersion) -> {
-            if (txVersion == null || txVersion >= trimMark) {
-                return;
-            }
-
-            conflictCache.invalidate(key);
-            entries.incrementAndGet();
-        });
-
-        log.info("Invalidated entries: {}", entries.get());
+        while(Address.isAddress(firstAddress()) && firstAddress() < trimMark) {
+            invalidateFirst();
+            entries++;
+        }
+        log.info("Invalidated entries {}", entries);
     }
 
     /**
      * The cache size
      *
      * @return cache size
      */
-    public long size() {
-        return conflictCache.estimatedSize();
+    public int size() {
+        return cacheConflictKeys.size();
     }
 
-    /**
+    /*
      * Put a value in the cache
      *
      * @param conflictStream conflict stream
-     * @param newTail        global tail
      */
-    public void put(ConflictTxStream conflictStream, long newTail) {
-        conflictCache.put(conflictStream, newTail);
-    }
+    public void put(ConflictTxStream conflictStream) {
+        if (cacheConflictKeys.size() == cacheSize) {
```

Comment:
Can you add a check that the conflict stream's version is greater than the old key's value (if it exists) and the corresponding tests. This cache implements a sliding window, but just based on these API's is possible to regress the key versions. Essentially, the put API always assumes that what ever key you put in, the caller has to guarantee that the corresponding value is greater,  but I think we should verify that regarderless of whether the caller does it or not.

Its a subtle point, but will catch any bugs in the sequencer server.